python/gps/algorithm/cost/tf_cost_utils.py

```diff
@@ -0,0 +1,446 @@
+import numpy as np
+import tensorflow as tf
+
+
+def assert_shape(tensor, shape):
+    assert tensor.get_shape().is_compatible_with(shape), "Shape mismatch: %s vs %s" % (str(tensor.get_shape()), shape)
+
+def logsumexp(x, reduction_indices=None):
+    """ Compute numerically stable logsumexp """
+    max_val = tf.reduce_max(x)
+    exp = tf.exp(x-max_val)
+    _partition = tf.reduce_sum(exp, reduction_indices=reduction_indices)
+    _log = tf.log(_partition)+max_val
+    return _log
+
+def safe_get(name, *args, **kwargs):
+    """ Same as tf.get_variable, except flips on reuse_variables automatically """
+    try:
+        return tf.get_variable(name, *args, **kwargs)
+    except ValueError:
+        tf.get_variable_scope().reuse_variables()
+        return tf.get_variable(name, *args, **kwargs)
+
+def init_weights(shape, name=None):
+    weights = np.random.normal(scale=0.01, size=shape).astype('f')
+    return safe_get(name, list(shape), initializer=tf.constant_initializer(weights))
+
+def init_bias(shape, name=None):
+    return safe_get(name, initializer=tf.zeros(shape, dtype='float'))
+
+def find_variable(name):
+    """ Find a trainable variable in a graph by its name (not including scope)
+
+    Example:
+    >>> find_variable('Wconv1')
+    <tensorflow tensor>
+    """
+    varnames = tf.trainable_variables()
+    matches = [varname for varname in varnames if varname.name.endswith(name+':0')]
+
+    if len(matches)>1:
+        raise ValueError('More than one variable with name %s. %s' % (name, [match.name for match in matches]))
+    if len(matches) == 0:
+        raise ValueError('No variables found with name %s. List: %s' % (name, [var.name for var in varnames]))
+    return matches[0]
+
+def jacobian(y, x):
+    """Compute derivative of y (vector) w.r.t. x (another vector)"""
+    dY = y.get_shape()[0].value
+    dX = x.get_shape()[0].value
+
+    deriv_list = []
+    for idx_y in range(dY):
+        grad = tf.gradients(y[idx_y], x)[0]
+        deriv_list.append(grad)
+    jac = tf.pack(deriv_list)
+    assert_shape(jac, [dY, dX])
+    return jac
+
+def multimodal_nn_cost_net_tf(num_hidden=3, dim_hidden=42, dim_input=27, T=100,
+                             demo_batch_size=5, sample_batch_size=5, phase=None, ioc_loss='ICML',
+                             Nq=1, smooth_reg_weight=0.0, mono_reg_weight=0.0, gp_reg_weight=0.0,
+                             multi_obj_supervised_wt=1.0, learn_wu=False, x_idx=None, img_idx=None,
+                              num_filters=[15,15,15]):
+    """ Construct cost net with images and robot config.
+    Args:
+        ...
+        x_idx is required, and should indicate the indices corresponding to the robot config
+        img_idx is required, and should indicate the indices corresponding to the imagej
+    """
+    inputs = {}
+    inputs['demo_obs'] = demo_obs = tf.placeholder(tf.float32, shape=(demo_batch_size, T, dim_input))
```

Comment:
a lot of the code here and in the below functions is copied. Can we reuse with helper functions?